File: solutions/ruby/spiral-matrix/1/spiral_matrix.rb
class SpiralMatrix
    attr_reader :matrix
    
    def initialize(dim)
        @matrix = Array.new(dim) { Array.new(dim) }
        startrow = startcol = 0
        endrow = endcol = dim - 1
        num = 1
        while startrow <= endrow && startcol <= endcol
            startcol.upto(endcol) do |i|
                @matrix[startrow][i] = num
                num += 1
            end
            startrow += 1
            startrow.upto(endrow) do |i|
                @matrix[i][endcol] = num
                num += 1
            end
            endcol -= 1
            endcol.downto(startcol) do |i|
                @matrix[endrow][i] = num
                num += 1
            end
            endrow -= 1
            endrow.downto(startrow) do |i|
                @matrix[i][startcol] = num
                num += 1
            end
            startcol += 1
        end
    end
end

File: solutions/typescript/spiral-matrix/1/spiral-matrix.ts
export function ofSize(dim: number): number[][] | [] {
  if (dim < 1) return [];
  // create a matrix of size dim x dim
  const matrix = Array.from({ length: dim }, () => Array(dim).fill(0));
  let counter = 1;
  let [startRow, startCol] = [0, 0];
  let endRow = dim - 1;
  let endCol = endRow;
  // populate a matrix in spiral order
  while (startRow <= endRow && startCol <= endCol) {
    // top row
    for (let i = startCol; i <= endCol; i++) {
      matrix[startRow][i] = counter++;
    }
    startRow++;
    // right column
    for (let i = startRow; i <= endRow; i++) {
      matrix[i][endCol] = counter++;
    }
    endCol--;
    // bottom row
    for (let i = endCol; i >= startCol; i--) {
      matrix[endRow][i] = counter++;
    }
    endRow--;
    // left column
    for (let i = endRow; i >= startRow; i--) {
      matrix[i][startCol] = counter++;
    }
    startCol++;
  }
  return matrix;
}
